vowpalwabbit/conditional_contextual_bandit.cc

```diff
@@ -9,15 +9,58 @@
 #include "cb_adf.h"
 #include "cb_algs.h"
 #include "constant.h"
+#include "object_pool.h"
 
 #include <numeric>
 #include <algorithm>
 #include <unordered_set>
+#include <queue>
 
 using namespace LEARNER;
 using namespace VW;
 using namespace VW::config;
 
+template <typename T>
+struct v_array_initializer
+{
+  v_array<T>* operator()(v_array<T>* array)
+  {
+    *array = v_array<T>();
+    return array;
+  }
+};
+
+template <typename T>
+struct v_array_cleanup
+{
+  void operator()(v_array<T>* array) { array->delete_v(); }
+};
+
+template <typename T>
+using v_array_pool = VW::no_lock_object_pool<v_array<T>, v_array_initializer<T>, v_array_cleanup<T>>;
+
+template <typename T>
+void return_v_array(v_array<T>& array, std::queue<v_array<T>*>& in_use_queue, v_array_pool<T>& pool)
+{
+  array.clear();
+  auto return_item = in_use_queue.front();
```

Comment:
Why do you need an in_use_queue?  Why not just pool.return_object(array)?

Reply:
The pool implementation being used requires that objects returned to the pool originated from the pool. If we did as you suggested it is invalid as it would be a copy of array and therefore a different object and address and not have originated from the pool.

I agree it is cumbersome, a different approach would be to accept returned objects by value and copy them into the pool which has it's own costs. And this is only valid for v_array because of its POD structure over a buffer.